src/common/selftest/selftest_esp_update.cpp

```diff
@@ -320,6 +224,145 @@ static void EspTask(void *pvParameters) {
     vTaskDelete(NULL); // kill itself
 }
 
+EspCredentials::EspCredentials(FSM_Holder &fsm, bool standalone)
+    : rfsm(fsm)
+    , standalone(standalone)
+    , progress_state(standalone ? esp_credential_action::ShowInstructions : esp_credential_action::CheckNeeded) {
+}
+
+bool EspCredentials::make_file() {
+    file.reset(fopen(file_name, "w"));
+    if (file.get() == nullptr) {
+        log_error(Network, "ESP credentials: Unable to create file %s", file_name);
+        return false;
+    }
+
+    fputs(file_str, file.get());
+    file.reset(nullptr);
+
+    log_info(Network, "ESP credentials generated to %s", file_name);
+    return true;
+}
+
+bool EspCredentials::file_exists() {
+    std::unique_ptr<FILE, FileDeleter> fl;
+
+    fl.reset(fopen(file_name, "r"));
+    return fl.get();
+}
+
+bool EspCredentials::upload_config() {
+    bool ret = netdev_load_esp_ini_to_eeprom();
+
+    notify_reconfigure();
+    return ret;
+}
+
+bool EspCredentials::already_set() const {
+    esp_entry_t entry = netdev_read_esp_ini_from_eeprom();
+    return (entry.ap.ssid && entry.ap.ssid[0] != '\0');
+}
+
+void EspCredentials::Loop() {
+    while (progress_state != esp_credential_action::Done) {
+        bool continue_pressed = false;
+
+        //we use only 3 responses here
+        //it is safe to use it from different thread as long as no other thread reads it
```

Comment:
`ClientResponseHandler::GetResponseFromPhase` is designed so It can be called from different thread. It works as long as only one thread at the time is using it (which is ensured here)